src/algebra/group/freiman.lean

```diff
@@ -0,0 +1,337 @@
+structure add_freiman_hom (α β : Type*) [add_comm_monoid α] [add_comm_monoid β] (n : ℕ) :=
+(to_fun : α → β)
+(map_sum_eq_map_sum' {s t : multiset α} (hs : s.card = n) (ht : t.card = n) (h : s.sum = t.sum) :
+  (s.map to_fun).sum = (t.map to_fun).sum)
+
+/-- A `n`-Freiman homomorphism is a map which preserves products of `n` elements. -/
+@[to_additive add_freiman_hom]
+structure freiman_hom (α β : Type*) [comm_monoid α] [comm_monoid β] (n : ℕ) :=
+(to_fun : α → β)
+(map_prod_eq_map_prod' {s t : multiset α} (hs : s.card = n) (ht : t.card = n)
+  (h : s.prod = t.prod) : (s.map to_fun).prod = (t.map to_fun).prod)
+
+notation α ` →+[`:25 n:25 `] `:0 β:0 := add_freiman_hom α β n
+notation α ` →*[`:25 n:25 `] `:0 β:0 := freiman_hom α β n
+
+/-- `add_freiman_hom_class F α β n` states that `F` is a type of `n`-ary sums-preserving morphisms.
+You should extend this class when you extend `add_freiman_hom`. -/
+class add_freiman_hom_class (F : Type*) (α β : out_param $ Type*) [add_comm_monoid α]
+  [add_comm_monoid β] (n : ℕ) extends fun_like F α (λ _, β) :=
+(map_sum_eq_map_sum' (f : F) {s t : multiset α} (hs : s.card = n) (ht : t.card = n)
+  (h : s.sum = t.sum) : (s.map f).sum = (t.map f).sum)
+
+/-- `freiman_hom_class F α β n` states that `F` is a type of `n`-ary products-preserving morphisms.
+You should extend this class when you extend `freiman_hom`. -/
+@[to_additive add_freiman_hom_class]
+class freiman_hom_class (F : Type*) (α β : out_param $ Type*) [comm_monoid α] [comm_monoid β]
+  (n : out_param ℕ)
+  extends fun_like F α (λ _, β) :=
+(map_prod_eq_map_prod' (f : F) {s t : multiset α} (hs : s.card = n) (ht : t.card = n)
+  (h : s.prod = t.prod) : (s.map f).prod = (t.map f).prod)
+
+-- `dangerous_instance` yields a false positive because `n` is an `out_param`.
+attribute [nolint dangerous_instance] freiman_hom_class.to_fun_like
+attribute [nolint dangerous_instance] add_freiman_hom_class.to_fun_like
+
+section comm_monoid
+variables [comm_monoid α] [comm_monoid β] [comm_monoid γ] [comm_monoid δ] [comm_group G] {n : ℕ}
+
+@[to_additive]
+lemma map_prod_eq_map_prod [freiman_hom_class F α β n] (f : F) {s t : multiset α} (hs : s.card = n)
+  (ht : t.card = n) (h : s.prod = t.prod) :
+  (s.map f).prod = (t.map f).prod :=
+freiman_hom_class.map_prod_eq_map_prod' f hs ht h
+
+namespace freiman_hom
+
+@[to_additive]
+instance freiman_hom_class : freiman_hom_class (α →*[n] β) α β n :=
+{ coe := to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_prod_eq_map_prod' := map_prod_eq_map_prod' }
+
+/-- Helper instance for when there's too many metavariables to apply `to_fun.to_coe_fn` directly. -/
+@[to_additive]
+instance : has_coe_to_fun (α →*[n] β) (λ _, α → β) := ⟨to_fun⟩
+
+@[to_additive]
+protected lemma map_prod (f : α →*[n] β) {s t : multiset α} (hs : s.card = n) (ht : t.card = n)
+  (h : s.prod = t.prod) :
+  (s.map f).prod = (t.map f).prod :=
+f.map_prod_eq_map_prod' hs ht h
+
```

Comment:
Could you try getting rid of this lemma altogether, replacing it with `map_prod_eq_map_prod`? The reason I kept around `whatever_hom.map_whatever` in #9888 is that it would take ages and ages to rename existing usages.